xtask/src/main.rs

```diff
@@ -0,0 +1,605 @@
+    out.contains("nightly")
+}
+
+fn load_expected_output(name: &str, release_mode: bool) -> Result<String> {
+    let path = Path::new("firmware/qemu/src/bin");
+
+    let filename;
+    if release_mode {
+        filename = format!("{}.release.out", name);
+    } else {
+        filename = format!("{}.out", name);
+    }
+
+    let path = path.join(filename);
+
+    // for error context closure
+    let path_str = path.to_str().unwrap_or("(non-Unicode path)").to_string();
+
+    let content = fs::read_to_string(path)
+        .with_context(|| format!("Failed to load expected output data from {}", path_str))?;
+    Ok(content)
+}
+
+fn test_single_snapshot(name: &str, features: &str, release_mode: bool) -> Result<()> {
+    let display_name = format!(
+        "{} ({})",
+        name,
+        if release_mode { "release" } else { "dev" }
+    );
+    println!("{}", display_name);
+    let cwd_name = "firmware/qemu".to_string();
+    let mut args;
+    if release_mode {
+        args = vec!["-q", "rrb", name]
+    } else {
+        args = vec!["-q", "rb", name]
+    }
+    if !features.is_empty() {
+        args.extend_from_slice(&["--features", features]);
+    }
+
+    let actual = run_capturing_stdout(Command::new("cargo").args(&args).current_dir(cwd_name))?;
+
+    let expected = load_expected_output(name, release_mode)?;
+
+    let diff = TextDiff::from_lines(&expected, &actual);
+
+    // if anything isn't ChangeTag::Equal, print it and turn on error flag
+    let mut actual_matches_expected = true;
+    for op in diff.ops() {
+        for change in diff.iter_changes(op) {
+            let styled_change = match change.tag() {
+                ChangeTag::Delete => Some(("-", Style::new().red())),
+                ChangeTag::Insert => Some(("+", Style::new().green())),
+                ChangeTag::Equal => None,
+            };
+            if let Some((sign, style)) = styled_change {
+                actual_matches_expected = false;
+                eprint!("{}{}", style.apply_to(sign).bold(), style.apply_to(change),);
+            }
+        }
+    }
+
+    if actual_matches_expected {
+        Ok(())
+    } else {
+        Err(anyhow!("{}", display_name))
+    }
+}
+
+fn get_installed_targets() -> Result<HashSet<String>> {
+    let out = run_capturing_stdout(Command::new("rustup").args(&["target", "list"]))?;
+    let installed_marker = " (installed)";
+    let mut targets = out.lines().collect::<Vec<_>>();
+    targets.retain(|target| target.contains(installed_marker));
+    let targets: HashSet<String> = targets
+        .iter()
+        .map(|target| target.replace(installed_marker, ""))
+        .collect();
+    Ok(targets)
+}
+
+fn install_targets() -> Result<Vec<String>> {
+    let required_targets = vec![
+        "thumbv6m-none-eabi",
+        "thumbv7m-none-eabi",
+        "thumbv7em-none-eabi",
+        "thumbv8m.base-none-eabi",
+        "riscv32i-unknown-none-elf",
+    ];
```

Comment:
make `required_targets` a `const [&str; 5]`?

Reply:
on another note: iirc there is some project providing macros to create hashmaps and hashsets ala `vec!`. maybe an idea for future readability improvement :)

Reply:
I prefer the flexibility of a `vec` (in case new targets come in) and consider the runtime overhead negligible. Again no super strong preference though.

Reply:
I don't really see why we need the runtime-flexibility of `Vec`. When adding targets we can just add it to the array, but you are right that this then would require to adapt the length in the type, when making it a `const`.

What do you think of:
```rust
let required_targets = [
    "thumbv6m-none-eabi",
    "thumbv7m-none-eabi",
    "thumbv7em-none-eabi",
    "thumbv8m.base-none-eabi",
    "riscv32i-unknown-none-elf",
]
.iter()
.map(|item| item.to_string())
.collect::<HashSet<_>>();

let installed_targets = get_installed_targets()?;
let added_targets = all_targets
    .difference(&installed_targets)
    .cloned()
    .collect::<Vec<_>>();
```
I realized that the intermediate `required_targets` isn't really needed, so we can merge the `all_targets` into it.

Reply:
`Vec` flexibility: I was strictly thinking in terms of code length (and hassle in case of changes), but it turns out I was having a brain fart moment when I thought a slice must be of compile time fixed size here. So yeah, let's get rid of superfluous `Vecs`!

Reply:
also, we're now one line shorter with your solution, you win 😆